src/gtk/textctrl.cpp

```diff
@@ -1030,111 +1030,21 @@ bool wxTextCtrl::IsEmpty() const
     return wxTextEntry::IsEmpty();
 }
 
-// Helper class used by DoSetValue() and WriteText() to temporarily reset
-// m_imKeyEvent to NULL and restore its original value in the dtor.
-class TemporarilyUnsetIMKeyEvent
-{
-public:
-    explicit TemporarilyUnsetIMKeyEvent(GdkEventKey*& event)
-        : m_event(event),
-          m_eventOrig(event)
-    {
-        m_event = NULL;
-    }
-
-    ~TemporarilyUnsetIMKeyEvent()
-    {
-        m_event = m_eventOrig;
-    }
-
-private:
-    GdkEventKey*& m_event;
-    GdkEventKey* const m_eventOrig;
-
-    wxDECLARE_NO_COPY_CLASS(TemporarilyUnsetIMKeyEvent);
-};
-
-void wxTextCtrl::DoSetValue( const wxString &value, int flags )
-{
-    wxCHECK_RET( m_text != NULL, wxT("invalid text ctrl") );
-
-    m_modified = false;
-
-    if ( !IsMultiLine() )
-    {
-        wxTextEntry::DoSetValue(value, flags);
-        return;
-    }
-
-    if (value.IsEmpty())
-    {
-        if ( !(flags & SetValue_SendEvent) )
-            EnableTextChangedEvents(false);
-
-        gtk_text_buffer_set_text( m_buffer, "", 0 );
-
-        if ( !(flags & SetValue_SendEvent) )
-            EnableTextChangedEvents(true);
-
-        return;
-    }
-
-#if wxUSE_UNICODE
-    const wxCharBuffer buffer(value.utf8_str());
-#else
-    wxFontEncoding enc = m_defaultStyle.HasFont()
-                            ? m_defaultStyle.GetFont().GetEncoding()
-                            : wxFONTENCODING_SYSTEM;
-    if ( enc == wxFONTENCODING_SYSTEM )
-        enc = GetTextEncoding();
-
-    const wxCharBuffer buffer(wxGTK_CONV_ENC(value, enc));
-    if ( !buffer )
-    {
-        // see comment in WriteText() as to why we must warn the user about
-        // this
-        wxLogWarning(_("Failed to set text in the text control."));
-        return;
-    }
-#endif
-
-    if ( !(flags & SetValue_SendEvent) )
-    {
-        EnableTextChangedEvents(false);
-    }
-
-    // This will emit insert-text signal which assumes that if m_imKeyEvent is
-    // set, it is called in response to this key press -- which is not the case
-    // here (but m_imKeyEvent might still be set e.g. because we're called from
-    // a menu event handler triggered by a keyboard accelerator).
-    TemporarilyUnsetIMKeyEvent unset(m_imKeyEvent);
-
-    gtk_text_buffer_set_text( m_buffer, buffer, strlen(buffer) );
-
-    if ( !(flags & SetValue_SendEvent) )
-    {
-        EnableTextChangedEvents(true);
-    }
-}
-
 void wxTextCtrl::WriteText( const wxString &text )
 {
     wxCHECK_RET( m_text != NULL, wxT("invalid text ctrl") );
 
     // we're changing the text programmatically
     DontMarkDirtyOnNextChange();
 
-    // avoid generating wxEVT_CHAR when called from wxEVT_CHAR handler
-    TemporarilyUnsetIMKeyEvent unset(m_imKeyEvent);
```

Comment:
It didn't seem to be necessary for single line controls in my testing of `SetValue()`, but it is indeed needed with `WriteValue()`, thanks for noticing this!